### ISSUE TITLE
Can't get RegEx fields working

### ISSUE THREAD

**Issue by brettwillis** [typesense/typesense#1091]
Sorry if this is a silly question, but I really can't seem to get RegEx fields working.

### Background
I have documents like this. Basically a set of fields on the document that are associated with any (variable/dynamic) field (tenant ID in this case).

```JSON
{
  ...
  "tenants": {
    "tenantA": {
      "balance": 0,
      ...
    },
    "tenantB": {
      "balance": 12.5,
      ...
    },
    "anyTenantId": {
      "balance": -3,
      ...
    },
  }
}
```
I want to be able to query by an exact defined (but dynamic) tenant ID. For examlpe, specifically tenant `tenantAbc123` (or any other specific ID): `?filter_by=tenants.tenantAbc123.balance:>0`, or `?sort_by=tenants.tenantDef456:desc`.

I can't flatten the `balance` fields into an array, because I loose the association with the tenant ID.

### What I've tried

With a schema like this:
```JSON
{ "name": "tenants..*.balance", type: "float" }
```

I always get `Cannot find a field named 'tenants.tenantAbc123.balance' in the schema`. I've also tried flattening the fields like `tenantAbc123_balance`, but same issue.

The only thing that I got to work was `{ "name": "tenants", type: "object" }` along with `enable_nested_fields: true`, but I run in to problems with type auto-detection, because it infers `int64` from initial integer values and then fails when it comes accross a `float` value. I know exactly the types of the fields, don't need auto-detection, I just need a wildcard in the field name but can't make it work?

### On a side note

The documentation about RegEx fields  rather sparse, it mentions "you can use any RegEx expression to define a field name", but is that really the case? Does it support anchors and character classes, alternations etc, or is it just `.*` (because that's the only example provided)? How are single dots (`.`) treated in a field name of a normal nested field (e.g. `nested_field.field1`)? Do they match any character and need to be escaped?

**Typesense Version**: 0.24.1

**Comment by kishorenc** [comment]
Some of the issues with auto detection of types inside nested fields are fixed in 0.25 RC builds (use `0.25.0.rc45`).  We have also added support for targeting nested field names using `.*` in 0.25 RC.

On the use of regular expressions in field names: you can use them for defining schema, but during querying, we currently only support the use of `.*` for nested fields and just `*` for targeting prefixes. Example:

```
query_by=name.*  -> can be used to search "name.first" and "name.last"
query_by=name*  -> can be used to search "name_first" and "name_last"
```

**Comment by brettwillis** [comment]
@kishorenc Ok I've given it a try with `0.25.0`, and have the following issues.

With a schema like this

```js
fields: [
  { name: 'tenants', type: 'object', optional: true },
  { name: 'tenants\\..*\\.balance', type: 'float', optional: true },
  { name: 'tenants\\..*\\.roles', type: 'string[]', optional: true },
  { name: 'tenants\\..*\\.updatedAt', type: 'string', optional: true, sort: true },
],
enable_nested_fields: true,
```

1. I can only use a concrete field like `tenants.tenantA.balance` at query time if `tenantA` actually exists in the data that has been imported at query time. Otherwise I get a "cannot find field for sorting" kind of error. As `tenants.tenantA.balance` matches a field in the schema (by wildcard) I would not expect that error, but it would behave as per usual when an optional field doesn't exist in the data.
2. If data has been imported with`tenantA` for example, then I can `sort_by=tenants.tenantA.balance:desc` (numeric field), but I cannot `sort_by=tenants.tenantA.updatedAt:desc` (string field) even though `sort: true` is specified on the field. I get the "cannot find field for sorting" error if it is a nested string field. Same with `filter_by`, it only works for numeric fields.
3. Even though `tenants\\..*\\.balance` is specified as `float`, if the first document encountered is an integer, then Typsense infers `int64` for that field. Then it truncates floating-point numbers to an integer when encountered in the following data.
4. The `tenants.{tenantId}` objects may contain a lot of other large data that I don't want to waste resources indexing. Having the { name: 'tenants', type: 'object' } field implies that all fields will be indexed, but none of the above examples work without it. Can we have a way to index *ONLY* the specified nested/wildcard fields, with the data type specified in the schema (not auto-detected)?

**Comment by kishorenc** [comment]
@brettwillis 

Thank you for putting together those limitations. We will be working to address them.

> Can we have a way to index ONLY the specified nested/wildcard fields (not every nested field), with the data type specified in the schema (not auto-detected)?

Have you tried not having the parent `{ name: 'tenants', type: 'object' }` definition?

**Comment by brettwillis** [comment]
> Have you tried not having the parent { name: 'tenants', type: 'object' } definition?

Yeah, if I don't have that definition, then using a concrete field name like in sorting or filtering like `sort_by=tenants.tenantA.balance:desc` always yeilds the "connot find field x in the schema for sorting" error. That is, using concrete field names in queries doesn't work in any way at all without that definition. 

Whereas with the definition, those queries do work, only for numeric fields, only when the field exists in the data.

I get the feeling that it is only the `{ name: 'tenants', type: 'object' }` that has any effect, and the other wildcard definitions actually have zero effect when I use concrete field names at query time?
